spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
 
-Towards Lite Client Accuracy:
+*Remark*: The variable *trustThreshold* can be used if the user believes that relying on one correct validator is not sufficient. However, in case of (frequent) changes in the validator set, the higher the *trustThreshold* is chosen, the more unlikely it becomes that CheckSupport returns true for non-adjacent headers.
 
-- Assume by contradiction that _h2_ was not generated correctly and the lite client sets trust to true because _CheckSupport_ returns true.
-- h1 is trusted and sufficiently new
-- by Tendermint Fault Model, less than 1/3 of voting power held by faulty validators => at least one correct validator _v_ has signed _h2_.
-- as _v_ is correct up to now, it followed the Tendermint consensus protocol at least up to signing _h2_ => _h2_ was correctly generated, we arrive at the required contradiction.
+**VerifyHeader.** The function *VerifyHeader* captures high level logic, i.e., application call to the lite client module to (optionally download) and
+verify header for some height. The core verification logic is captured by *CanTrust* function that iteratively try to establish trust in given header
+by relying on *CheckSupport* function.
 
-Towards Lite Client Completeness:
 
-- The check is successful if sufficiently many validators of _h1_ are still validators in _h2_ and signed _h2_.
-- If _h2.Header.height = h1.Header.height + 1_, and both headers were generated correctly, the test passes
+```go
+func VerifyHeader(height, trustThreshold) error {
+  if h2, exists := Store.Get(height); exists {
+    if isWithinTrustedPeriod(h2) { return nil }
+    return ErrHeaderNotWithinTrustedPeriod(h2)
+  }
 
-_Verification Condition:_ We may need a Tendermint invariant stating that if _h2.Header.height = h1.Header.height + 1_ then _signers(h2.Commit) \subseteq h1.Header.NextV_.
+  h2 := Commit(height)
+  if !verify(h2) { return ErrInvalidHeader(h2) }
+  if !isWithinTrustedPeriod(h2) { return ErrHeaderNotWithinTrustedPeriod(h2) }
 
-_Remark_: The variable _trustlevel_ can be used if the user believes that relying on one correct validator is not sufficient. However, in case of (frequent) changes in the validator set, the higher the _trustlevel_ is chosen, the more unlikely it becomes that CheckSupport returns true for non-adjacent headers.
+  // get the highest trusted headers lower than h2
+  h1 = Store.HighestTrustedSmallerThan(height)
+  if h1 == nil { return ErrNoTrustedHeader }
 
-**Bisection.** The following function uses CheckSupport in a recursion to find intermediate headers that allow to establish a sequence of trust.
+  err = CanTrust(h1, h2, trustThreshold)  // or CanTrustBisection((h1, h2, trustThreshold)
+  if err != nil { return err }
 
-```go
-func Bisection(h1,h2,trustlevel) bool{
-  if CheckSupport(h1,h2,trustlevel) {
-    return true
+  if isWithinTrustedPeriod(h2) {
+    Store.add(h2)
```

Comment:
Isn't this redundant? `CanTrust` already has stored `h2` here. At least in the bisection case because:
```go
func CanTrustBisection(h1,h2,trustThreshold) error {
  assume h1.Header.Height < h2.header.Height

  err = CheckSupport(h1,h2,trustThreshold)
  if err == nil {
    Store.Add(h2)
    return nil
  }
  if err != ErrTooMuchChange return err

  pivot := (h1.Header.height + h2.Header.height) / 2
  hp := Commit(pivot)
  if !verify(hp) return ErrInvalidHeader(hp)

  err = CanTrustBisection(h1,hp,trustThreshold)
  if err == nil {
      Store.Add(hp)
      err2 = CanTrustBisection(hp,h2,trustThreshold)
      if err2 == nil {
        Store.Add(h2)
        return nil
      }
      return err2
  }
  return err
}
```

Reply:
Bisection needs to store headers in between (the pivot headers). But maybe we leave it to the caller to actually deal with the last header `h2`? 

I think the name of this method (`CanTrust` / `CanTrustBisection`) should actually indicate that it is updating the store and not just answering the question "Can we trust this header based on h1".